libs/util/AdvancedMath.arucas

```
@@ -0,0 +1,915 @@
+    var imaginary;
+    var arg;
+    var abs;
+
+    //"Constructors"
+    ComplexNumber(r : Number, i : Number){
+        this.real = r;
+        this.imaginary = i;
+        this.arg = Math.arctan(i/r);
+        this.abs = Math.sqrt(r*r + i*i);
+    }
+
+    ComplexNumber(polar : List){
+        this.abs = polar[0];
+        this.arg = polar[1];
+        this.real = this.abs * Math.cos(this.arg);
+        this.imaginary = this.abs * Math.sin(this.arg);
+    }
+
+    static fun complex(r : Number, i : Number): ComplexNumber{
+        return new ComplexNumber(r,i);
+    }
+
+
+    // Operators
+    operator +(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.add(other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.add(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator -(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.substract(other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.substract(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator *(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.multiply(other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.multiply(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator /(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.multiply(1/other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.divide(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator ^(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.exponent(other);
+        }
+        else{
+            return null;
```

Should you really be returning null here? I feel like that will just lead to confusion, I would throw an error, or even better type hint the parameter in which case it will guarentee the type you can even allow multiple types like `fun s(other: Number | ComplexNumber)`, this applies to a few of the methods here.

---

I meant to comment on the parameters, but apparently I clicked wrong

---

delegated this to surf by tagging on discord